addons/api/addon/handlers/indexed-db-handler.js

```diff
@@ -0,0 +1,61 @@
+import { modelIndexes } from 'api/services/indexed-db';
+
+/**
+ * Handler to sit in front of the API request layer
+ * so we can handle any caching with indexedDb first
+ */
+const IndexedDbHandler = {
+  async request(context, next) {
+    switch (context.request.op) {
+      case 'query': {
+        const { store, data } = context.request;
+        const { type, query } = data;
+        const supportedModels = Object.keys(modelIndexes);
+        const { db: indexedDb } = this.indexedDb ?? {};
+
+        // TODO: Differentiate between user initiated paginated
+        //  requests and cache population requests
+
+        // Go through normal flow if we don't yet support the model
+        // or if we don't have an indexedDb instance
+        if (!supportedModels.includes(type) || !indexedDb) {
+          // console.log(this.indexedDb.db);
+          break;
+        }
+
+        const adapter = store.adapterFor(type);
+        const schema = store.modelFor(type);
+
+        const payload = await adapter.query(store, schema, query);
+        const serializer = store.serializerFor(type);
+        const normalizedPayload = serializer.normalizeResponse(
+          store,
+          schema,
+          payload,
+          null,
+          'query'
+        );
+
+        // eslint-disable-next-line no-unused-vars
+        const { data: payloadData } = normalizedPayload;
+
+        // TODO: Just to show how to add data to indexedDb
+        // await this.indexedDb.db[type].bulkPut(payloadData);
+        // await this.indexedDb.db[type]
+        //   .where('attributes.type')
+        //   .equals('ssh')
+        //   .toArray();
+
+        // TODO: Manually insert the normalized data into ember data store and don't call next()
```

Comment:
Non-blocking question for understanding: once this TODO is completed does that mean  that the `LegacyNetworkHandler` use will be removed ?

Reply:
No, we will have `LegacyNetworkHandler` for the forseeable future. That handles everything for how the old style adapters and serializers work. If you notice, I only hook in currently for `query` requests and even then only for very specific scenarios, otherwise every other request will go through the legacy handler.